htp: remove server bypass, add early response detection

sirthias opened the pull request:
Closes #16898 and solves the first (server-side) half of #18830.

/cc @jrudolph

sirthias committed:
=htp #18830 simplify HttpServerBluePrint: removing bypass, clean up 100-continue support

sirthias committed:
=htp #16898 detect early responses, warn + close if detected

ktoso commented:
Awesome, thanks - will review soon, out of home now. Have a nice weekend in any case! :)

jrudolph commented:
Nice!

sirthias commented:
One benefit of this refactoring: The server-layer is now, finally, fully reusable!

drewhk commented:
Can we actually cache it then in the relevant places?

sirthias commented:
Usually it is not really reused anywhere. You only benefit from the reusability if you bind, unbind, and rebind, which should be a rare use case.

jrudolph commented:
Actually, the websocket part is probably not reusable.

sirthias commented:
Ah, ok, that sucks. I thought the ugly hack involving the ActorRef to the OneHundredContinue trigger was all that prevented reusability.

analytically commented:
Impressive work!

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +87,222 @@ private[http] object HttpServerBluePrint {
+        csRequestPrepOut ~> requestPrep
+
+        // One2OneBidi
+        val one2one = b.add(new One2OneBidi[HttpRequest, HttpResponse](16)) // TODO: replace hard-coded value by reintroducing `pipeline-limit` setting from spray
```
Could you try to address this TODO as part of this PR already @sirthias?

sirthias replied:
Ok, will do.

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +87,222 @@ private[http] object HttpServerBluePrint {
-      setHandler(bypassInput, new InHandler {
-        override def onPush(): Unit = {
-          grab(bypassInput) match {
+  class ControllerStage(settings: ServerSettings, log: LoggingAdapter)
```
reads much nicer indeed :+1:

drewhk commented:
Please enable `akka.stream.materializer.debug.fuzzing-mode=on` on the relevant akka-http projects to see if these changes cause any problems. Probably another ticket.

ktoso commented:
Failure was unrelated (AcknowledgeSink for which we have a ticket), LGTM.
Would you like to amend the remaining TODO before we merge here @sirthias ?

sirthias commented:
@drewhk You mean change the config in all tests?

drewhk commented:
you should put it into /test/resources/reference.conf so that it applies to all tests and then disable it in tests where it is provably wrong to enable it.

sirthias commented:
@drewhk I see, thanks. Will do.

sirthias committed:
=htc,htp prepare for enabling `stream.materializer.debug.fuzzing-mode=on` in tests

sirthias commented:
@drewhk With `fuzzing-mode=on` I get sporadic future timeout errors in 2 specs of the http-core tests in about 50% of the cases (i.e. about half the time all http-core tests pass on my machine).
So, there does seem to be something to investigate. I'll therefore create a separate ticket for enabling fuzzing-mode in http tests and investigate errors, so we can get this thing here merged first.

drewhk commented:
Ok good! It is a quite cool feature, it would be nice to have it enabled everywhere :)

sirthias committed:
+htc migrate `pipelining-limit` server config setting from spray

sirthias commented:
Ok, rebased, added `pipelining-limit` server config setting and prepared for enabling fuzzing mode.
Also added a pretty ascii art showing the server-side HTTP pipeline setup.

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
 
-    BidiFlow.fromGraph(FlowGraph.create(requestParsingFlow, rendererPipeline, oneHundredContinueSource)((_, _, _) ⇒ ()) { implicit b ⇒
-      (requestParsing, renderer, oneHundreds) ⇒
+    BidiFlow.fromGraph(FlowGraph.create(requestParsingFlow, rendererPipeline)(Keep.none) { implicit b ⇒
```
passing parameters here is only for the purpose of using their materialized values—it would seem that requestParsing and rendererPipeline might as well be imported like all the others below

drewhk replied:
+1

sirthias replied:
I see. I was wondering about the difference myself.
Thanks for the clarification, will simplify.

drewhk commented:
@sirthias @jrudolph can you tell which parts are still not reusable? We might be able to fix those now or very soon, but I don't know what is the current non-reusable part and why.

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
-      case m: MessageStartError ⇒ m
-    }
-
     val rendererPipeline =
```
minor nitpick: rename `requestParsingFlow`/`rendererPipeline` to `requestParsing`/`responseRendering`?

sirthias replied:
Yup, that's better!

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
+        switchTokenBroadcast ~> switchToWebsocket
+        protocolRouter.out0 ~> http ~> switchTokenBroadcast ~> protocolMerge.in0
+        protocolRouter.out1 ~> websocket ~> protocolMerge.in1
+        switchToWebsocket ~> wsSwitchTokenMerge.in1
```
please fold into three lines above: `switchTokenBroadcast ~> switchToWebsocket ~> wsSwitchTokenMerge.in1`

sirthias replied:
Yes, will do.

sirthias commented:
@drewhk The HTTP part itself should be fully reusable.
@jrudolph said that there might be non-reusable stuff in the websocket graph, but I don't know/remember which things that comprises.

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
-      def finishWithError(status: StatusCode, info: ErrorInfo): Unit = {
+      setHandler(responseCtxOut, new OutHandler {
+        def onPull(): Unit = if (!hasBeenPulled(httpResponseIn)) pull(httpResponseIn)
+        override def onDownstreamFinish() = cancel(httpResponseIn)
```
if there are outstanding requests then this will deadlock the stage, right?

sirthias replied:
I'd expect the cancellation to propagate through the application and the incoming side back to the TCP layer.
But I guess we could also `completeStage()` here since it's unlikely that there is still interesting stuff going on on the incoming side.

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
         }
+        override def onUpstreamFinish() =
+          if (openRequests.isEmpty && isClosed(requestParsingIn)) completeStage()
+          else complete(responseCtxOut)
```
there won’t be any more responses, so why not complete the stage right away?

sirthias replied:
Because in certain cases (involving early responses!) the *entity* of the last response might depend on the entity of the last request still being read from the network. A simple example would be the classic "ping/pong"  where you stuff the request entity directly into the response. We actually have a test for exactly this setup.
If we complete the stage here we kill the incoming side as well and therefore the entity of the last response.

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
+          val response = grab(httpResponseIn)
+          val requestStart = openRequests.head
+          openRequests = openRequests.tail
+          val isEarlyResponse = messageEndPending && openRequests.isEmpty
```
this means that in the face of pipelining early responses are not detected; fixing this would be highly welcome :-) (since we’ll want to piggy-back on the code in order to consume non-materialized request entities at this point as well)

sirthias replied:
If we already have another request queued then the response cannot be an "early" one anymore.
The `messageEndPending` here always concerns the last request in the queue. For all earlier ones we must have already seen the end of the request entity, otherwise we wouldn't have been able to read (and queue) the next request!

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
 
-      def finishWithError(status: StatusCode, info: ErrorInfo): Unit = {
+      setHandler(responseCtxOut, new OutHandler {
+        def onPull(): Unit = if (!hasBeenPulled(httpResponseIn)) pull(httpResponseIn)
```
This is fishy (in general—using hasBeenPulled reeks of not knowing what state we’re in); in particular it is quite hard to predict when this handler will be in effect due to the simultaneous use of `emit`.

sirthias replied:
True. Thinking about it again there is currently potential for a "push-without-preceding-pull" error.
I'll straighten this out.
Thanks for the cue!

drewhk replied:
No worries, once you enable the fuzzer it will find many of these ;) (talking from experience)

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -96,208 +105,222 @@ private[http] object HttpServerBluePrint {
+            requestStart.expect100Continue && oneHundredContinueResponsePending ||
+            isClosed(requestParsingIn) && openRequests.isEmpty ||
+            isEarlyResponse
+          push(responseCtxOut, ResponseRenderingContext(response, requestStart.method, requestStart.protocol, close))
```
I cannot prove that this is safe here; my recommendation is to `emit` exclusively on this output port, and add a `() => pull(httpResponseIn)` (lifted into a `val`, of course)

sirthias replied:
Yes, thanks!

sirthias replied:
Coming back to this:
Wouldn't an `emit(..., () => pull(httpResponseIn))` create a potential OOME here?
After all we'd be pulling from the upstream at full speed without any backpressure from downstream allowing to come in!

rkuhn replied:
No, that is precisely what the callback is for: it is only executed once we have emitted to the Outlet, not before.

sirthias replied:
Ok, I see. Thx!

jrudolph commented:
I think the only thing that isn't reusable is the `websocketSetup`. To make that reusable you would have to access materialized values of the Websocket SubscriberSource / PublisherSink stubs during materialization.

rkuhn commented on akka-stream/src/main/scala/akka/stream/scaladsl/One2OneBidiFlow.scala:
```diff
@@ -77,10 +75,7 @@ object One2OneBidiFlow {
-          cancel(outIn)
-          cancel(inIn) // short-cut to speed up cleanup of upstream
-        }
+        override def onDownstreamFinish(): Unit = cancel(outIn)
```
point 4) above is not true, it will signal UnexpectedOutputException in this case

`GraphStageLogic.complete(...)` should get the same treatment of Emitting states that `completeStage()` has, I suspect that this may be the cause of some problems you see

`pullsSuppressed` can only ever be 0 or 1, i.e. it should be a boolean

sirthias replied:
> point 4) above is not true, it will signal UnexpectedOutputException in this case

Yes, true. Appears to be a left-over from an earlier implementation.
I'll simply remove point 4) from the scaladoc.

> pullsSuppressed can only ever be 0 or 1, i.e. it should be a boolean

Yes. I always keep forgetting that there can only every be one pull and one push pending.

> GraphStageLogic.complete(...) should get the same treatment of Emitting states that completeStage() has, I suspect that this may be the cause of some problems you see

I don't understand what you mean by that. Can you clarify?

rkuhn replied:
I mean that it should check the currently installed handler and see if it is Emitting, scheduling the completion for later in this case (see https://github.com/akka/akka/blob/release-2.3-dev/akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala#L393).

sirthias replied:
Ah, yes, that makes sense.
Should I make the change here or postpone to an extra ticket?

rkuhn replied:
Please make it here, it might be that this PR is affected by the lack of this feature.

drewhk replied:
> Yes. I always keep forgetting that there can only every be one pull and one push pending.

`PushPullStage` enforces this mostly at compile time with proper types, but `GraphStage` does not. Btw this simple invariant is what allows us to make these interpreters allocation free.

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -29,6 +29,24 @@ import akka.util.ByteString
+ * |                  HttpResponse                     |             |     Response-     | renderer- |     ByteString
+ * v -------------------------------------------------->             +-------------------> Pipeline  +-------------->
+ *                                                     |             |     Rendering-    |           |
+ *                                                     +-------------+     Context       +-----------+
```
I know it is getting big, but not including the One2One stage, websockets and TLS is slightly confusing; maybe there should at least be some indication that something is missing.

sirthias committed:
=str,htc apply review feedback: smaller improvement across the board

sirthias commented:
Ok, applied all review feedback so far.
Ready for next round.

sirthias commented:
@ktoso Does this make it into M2 as well?

drewhk commented:
I would like to. Any objections?

ktoso commented:
Yes let's include it as well

ktoso commented:
I'll give the last commits a look now before merging.

ktoso commented:
Last commits look good, thanks @sirthias

jrudolph commented:
I just figured that passing the RemoteAddress to the request parser is another reason the serverblueprint is not reusable. I created #19050 to track the effort.
